From 9a90b3ec0a21c72520c6031292fa809104f9917c Mon Sep 17 00:00:00 2001
From: Wang Xuerui <idontknw.wang@gmail.com>
Date: Mon, 13 Jun 2016 17:09:34 +0800
Subject: [PATCH] cat: minimally port MTK CC Alpha msg handling

Without this, com.android.phone will FC with ClassCastException when
making calls.

Change-Id: Id8466cfa7ba2e31a5c62863236490876a34c280a
---
 .../internal/telephony/cat/AppInterface.java       |  7 +++-
 .../internal/telephony/cat/CatCmdMessage.java      | 10 +++++
 .../android/internal/telephony/cat/CatService.java | 43 +++++++++++++++++++++-
 .../internal/telephony/cat/CommandParams.java      | 19 ++++++++++
 4 files changed, 77 insertions(+), 2 deletions(-)

diff --git a/src/java/com/android/internal/telephony/cat/AppInterface.java b/src/java/com/android/internal/telephony/cat/AppInterface.java
index 681f069f8d..6f16f73879 100644
--- a/src/java/com/android/internal/telephony/cat/AppInterface.java
+++ b/src/java/com/android/internal/telephony/cat/AppInterface.java
@@ -94,7 +94,12 @@
         RECEIVE_DATA(0x42),
         SEND_DATA(0x43),
         GET_CHANNEL_STATUS(0x44),
-        ACTIVATE(0x70);
+        ACTIVATE(0x70),
+        // MTK
+        /**
+         * Proprietay message for Call Control alpha id display
+         */
+        CALLCTRL_RSP_MSG(0XFF);
 
         private int mValue;
 
diff --git a/src/java/com/android/internal/telephony/cat/CatCmdMessage.java b/src/java/com/android/internal/telephony/cat/CatCmdMessage.java
index e99473c213..0fa62e3452 100644
--- a/src/java/com/android/internal/telephony/cat/CatCmdMessage.java
+++ b/src/java/com/android/internal/telephony/cat/CatCmdMessage.java
@@ -35,6 +35,9 @@
     private CallSettings mCallSettings = null;
     private SetupEventListSettings mSetupEventListSettings = null;
     private boolean mLoadIconFailed = false;
+    // MTK
+    public int mInfoType = 0;
+    public String mDestAddress = null;
 
     // Command Qualifier values for refresh command
     static final int REFRESH_NAA_INIT_AND_FULL_FILE_CHANGE  = 0x00;
@@ -129,6 +132,13 @@
             break;
         case ACTIVATE:
         case PROVIDE_LOCAL_INFORMATION:
+            break;
+        // MTK
+        case CALLCTRL_RSP_MSG:
+            mTextMsg = ((CallCtrlBySimParams) cmdParams).mTextMsg;
+            mInfoType = ((CallCtrlBySimParams) cmdParams).mInfoType;
+            mDestAddress = ((CallCtrlBySimParams) cmdParams).mDestAddress;
+            break;
         default:
             break;
         }
diff --git a/src/java/com/android/internal/telephony/cat/CatService.java b/src/java/com/android/internal/telephony/cat/CatService.java
index 9151c7572b..fb25fc1cca 100755
--- a/src/java/com/android/internal/telephony/cat/CatService.java
+++ b/src/java/com/android/internal/telephony/cat/CatService.java
@@ -891,11 +891,52 @@ public void handleMessage(Message msg) {
             }
             break;
         case MSG_ID_ALPHA_NOTIFY:
+            // MTK
             CatLog.d(this, "Received CAT CC Alpha message from card");
             if (msg.obj != null) {
                 AsyncResult ar = (AsyncResult) msg.obj;
                 if (ar != null && ar.result != null) {
-                    broadcastAlphaMessage((String)ar.result);
+                    if (ar.result instanceof String) {
+                        broadcastAlphaMessage((String)ar.result);
+                    } else if (ar.result instanceof String[]) {
+                        // MTK
+                        String[] callCtrlInfo = (String[]) ar.result;
+                        CatLog.d(this, "callCtrlInfo.length: " + callCtrlInfo.length + "," +
+                                callCtrlInfo[0] + "," + callCtrlInfo[1] + "," +
+                                callCtrlInfo[2]);
+                        byte[] rawData = null;
+                        try {
+                            if (null != callCtrlInfo[1] && callCtrlInfo[1].length() > 0) {
+                                rawData = IccUtils.hexStringToBytes(callCtrlInfo[1]);
+                            } else {
+                                CatLog.d(this, "Null CC alpha id.");
+                                break;
+                            }
+                        } catch (Exception e) {
+                            // zombie messages are dropped
+                            CatLog.d(this, "CC message drop");
+                            break;
+                        }
+                        String alphaId = null;
+                        try {
+                            alphaId = IccUtils.adnStringFieldToString(
+                                    rawData, 0, rawData.length);
+                        } catch (IndexOutOfBoundsException e) {
+                            CatLog.d(this, "IndexOutOfBoundsException adnStringFieldToString");
+                            break;
+                        }
+                        CatLog.d(this, "CC Alpha msg: " + alphaId + ", sim id: " + mSlotId);
+                        TextMessage textMessage = new TextMessage();
+                        CommandDetails cmdDet = new CommandDetails();
+                        cmdDet.typeOfCommand = AppInterface.CommandType.CALLCTRL_RSP_MSG.value();
+                        textMessage.text = alphaId;
+                        CallCtrlBySimParams cmdParams = new CallCtrlBySimParams(cmdDet,
+                                textMessage, Integer.parseInt(callCtrlInfo[0]), callCtrlInfo[2]);
+                        CatCmdMessage cmdMsg = new CatCmdMessage(cmdParams);
+                        broadcastCatCmdIntent(cmdMsg);
+                    } else {
+                        CatLog.d(this, "CAT Alpha message: unexpected response type!");
+                    }
                 } else {
                     CatLog.d(this, "CAT Alpha message: ar.result is null");
                 }
diff --git a/src/java/com/android/internal/telephony/cat/CommandParams.java b/src/java/com/android/internal/telephony/cat/CommandParams.java
index b87b042be0..c6df26c96c 100644
--- a/src/java/com/android/internal/telephony/cat/CommandParams.java
+++ b/src/java/com/android/internal/telephony/cat/CommandParams.java
@@ -254,3 +254,22 @@ boolean setIcon(Bitmap icon) {
         ussdString = ussdstring;
     }
 }
+
+// MTK
+
+/**
+ * CallCtrlBySimParams class for CC by SIM parameters.
+ */
+class CallCtrlBySimParams extends CommandParams {
+    TextMessage mTextMsg;
+    int mInfoType;
+    String mDestAddress;
+
+    CallCtrlBySimParams(CommandDetails cmdDet, TextMessage textMsg,
+        int infoType, String destAddress) {
+        super(cmdDet);
+        mTextMsg = textMsg;
+        mInfoType = infoType;
+        mDestAddress = destAddress;
+    }
+}
